pkg/trie/triedb/recorder.go

```diff
@@ -0,0 +1,93 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package triedb
+
+import (
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/tidwall/btree"
+)
+
+type TrieAccess interface {
+	isTrieAccess()
+}
+
+type (
+	EncodedNodeAccess struct {
+		hash        common.Hash
+		encodedNode []byte
+	}
+	ValueAccess struct {
+		// We are not using common.Hash here since hash size could be > 32 bytes when we use prefixed keys
+		hash    []byte
+		value   []byte
+		fullKey []byte
+	}
+	InlineValueAccess struct {
+		fullKey []byte
+	}
+	HashAccess struct {
+		fullKey []byte
+	}
+	NonExistingNodeAccess struct {
+		fullKey []byte
+	}
+)
+
+func (EncodedNodeAccess) isTrieAccess()     {}
+func (ValueAccess) isTrieAccess()           {}
+func (InlineValueAccess) isTrieAccess()     {}
+func (HashAccess) isTrieAccess()            {}
+func (NonExistingNodeAccess) isTrieAccess() {}
+
+type RecordedForKey int
+
+const (
+	RecordedValue RecordedForKey = iota
+	RecordedHash
+)
+
+type Record struct {
+	// We are not using common.Hash here since hash size could be > 32 bytes when we use prefixed keys.
+	// See ValueAccess.hash
+	hash []byte
+	data []byte
+}
+
+type Recorder struct {
+	nodes        []Record
+	recordedKeys btree.Map[string, RecordedForKey]
+}
```

Comment:
Same for these, do they need to be public?

Reply:
Yes, since they are going to be used by the proof generation in a different package